File: ProjectTracker20vb/ProjectTracker.Library/ProjectInfo.vb
<Serializable()> _
Public Class ProjectInfo
  Inherits ReadOnlyBase(Of ProjectInfo)

  Private mId As Guid
  Private mName As String

  Public Property Id() As Guid
    Get
      Return mId
    End Get
    Friend Set(ByVal value As Guid)
      mId = value
    End Set
  End Property

  Public Property Name() As String
    Get
      Return mName
    End Get
    Friend Set(ByVal value As String)
      mName = value
    End Set
  End Property

  Protected Overrides Function GetIdValue() As Object
    Return mId
  End Function

  Public Overrides Function ToString() As String
    Return mName
  End Function

  Private Sub New()
    ' require use of factory methods
  End Sub

  Friend Sub New(ByVal id As Guid, ByVal name As String)
    mId = id
    mName = name
  End Sub

End Class

File: ProjectTrackercs/ProjectTracker.Library/ProjectInfo.cs
using System;
using Csla;
using Csla.Serialization;

namespace ProjectTracker.Library
{
  [Serializable()]
  public partial class ProjectInfo : ReadOnlyBase<ProjectInfo>
  {
    private Guid _id;
    private string _name;

    public Guid Id
    {
      get
      {
        return _id;
      }
      internal set
      {
        _id = value;
      }
    }

    public string Name
    {
      get
      {
        return _name;
      }
      internal set
      {
        _name = value;
      }
    }

    protected override object GetIdValue()
    {
      return _id;
    }

    public override string ToString()
    {
      return _name;
    }

    private ProjectInfo()
    {
      // require use of factory methods
    }

    internal ProjectInfo(Guid id, string name)
    {
      _id = id;
      _name = name;
    }
  }
}